### PR TITLE
feat(banner): banner without title and no clickable

### PR DESCRIPTION
**Pull request by christianpe98**
## Motivation
From motive, our intention is to have banners that are not clickable and do not have a title. Right now, the X banner forces any banner to have a URL to redirect to and a title. In this PR we propose a solution to support X banners and the banners Motive needs.

### `banner.model.ts`

- `url` and `title` are optional.

### `banner.vue`

- If the banner has a URL it will be clickable, triggering the appropriate events, otherwise it will just be an image(with or without title). 
- If there is an error loading the banner image, the banner will be removed from the DOM.
- If the banner does not have a title, the alternative text of the image will be "Banner" and the heading(`h2`) will not be rendered.

### `banner.spec.ts`
We have defined some unit tests at component level using Jest(`x-components/src/x-modules/search/components/__tests__/banner.spec.ts`) and others using Cypress(`x-components/tests/unit/banner.spec.ts`). We use Cypress to check that when loading an image, if an error occurs the banner is not rendered.

For the tests we have modified the existing banner factory, this has caused some changes in other tests(`mocked-responses.spec.ts`).

### CODE REVIEW

File: packages/x-components/src/x-modules/search/components/banner.vue [hunk 1]
```diff
@@ -1,15 +1,30 @@
...
+    <h2 v-if="banner.title" class="x-banner__title">{{ banner.title }}</h2>
   </a>
+  <figure v-else-if="!imageError" class="x-banner" data-test="banner">
+    <img
+      @error="imageError = true"
+      :src="banner.image"
+      :alt="banner.title ? banner.title : 'Banner'"
+      class="x-banner__image"
+    />
+    <h2 v-if="banner.title" class="x-banner__title">{{ banner.title }}</h2>
```
Comment (CachedaCodes) [inline packages/x-components/src/x-modules/search/components/banner.vue:26]: This is duplicated. Try to find a solution that doesn't have this duplicity.

Changing the root element to a conditional `component` element and making adding the events dynamic could work:

````
<component v-if="!imageError" :is="banner.url ? 'a' : 'figure'"
    v-if="!imageError"
    v-on="banner.url ? { // don't set this object here
      click: emitClickEvent,
      auxclick: emitClickEvent,
      contextmenu: emitClickEvent
    } : {}"
    :href="banner.url"
    class="x-banner"
    data-test="banner"
  >
    <img
      @error="imageError = true"
      :src="banner.image"
      :alt="banner.title ? banner.title : 'Banner'"
      class="x-banner__image"
    />
    <h2 v-if="banner.title" class="x-banner__title">{{ banner.title }}</h2>
  </component>
````

Other solution could be moving the click anchor and logic to a component and the do:

````
<component v-if="!imageError" :is="banner.url ? 'newComponent' : 'figure'"
    v-if="!imageError"
    class="x-banner"
    data-test="banner"
  >[...]
````

File: packages/x-components/src/x-modules/search/components/banner.vue [hunk 2]
```diff
@@ -1,15 +1,30 @@
 <template>
   <a
+    v-if="banner.url && !imageError"
```
Comment (CachedaCodes) [inline packages/x-components/src/x-modules/search/components/banner.vue:3]: Now that this is handling the case where the image load errors, there should be a test for it.

**Review by CachedaCodes — CHANGES_REQUESTED** [review]

File: packages/x-components/src/x-modules/search/components/banner.vue [hunk 3]
```diff
@@ -38,6 +53,13 @@
     @Prop({ required: true })
     public banner!: BannerModel;
 
+    /**
+     * Flag to handle banner image errors.
+     *
+     * @public
+     */
+    protected imageError = false;
```
[the remark this answers is not in the document]
Comment (christianpe98) [reply packages/x-components/src/x-modules/search/components/banner.vue]: done

File: packages/x-components/src/x-modules/search/components/banner.vue (hunk 2 above)
Comment (christianpe98) [reply packages/x-components/src/x-modules/search/components/banner.vue]: done

File: packages/x-components/src/x-modules/search/components/banner.vue (hunk 1 above)
Comment (christianpe98) [reply packages/x-components/src/x-modules/search/components/banner.vue]: done

File: packages/x-components/src/x-modules/search/components/__tests__/banner.spec.ts [hunk 4]
```diff
@@ -44,28 +44,26 @@ describe('testing Banner component', () => {
...
+    const { wrapper } = renderBanner({
+      banner: createBannerStub('banner')
+    });
+
+    expect(wrapper.get(getDataTestSelector('banner')).text()).toEqual('');
+  });
+
   // eslint-disable-next-line max-len
-  it('emits UserClickedABanner when the user clicks in the left, middle or right button on the component', () => {
+  it('renders a banner which emits UserClickedABanner when the user clicks in the left, middle or right button on the component', () => {
```
Comment (CachedaCodes) [inline packages/x-components/src/x-modules/search/components/__tests__/banner.spec.ts:64]: This test fails, check it again after changing the implementation of the stubs.

**Review by CachedaCodes — CHANGES_REQUESTED** [review]

File: packages/x-components/src/x-modules/search/components/__tests__/banner.spec.ts (hunk 4 above)
Comment (christianpe98) [reply packages/x-components/src/x-modules/search/components/__tests__/banner.spec.ts:64]: done

File: packages/x-components/tests/unit/banner.spec.ts [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      return cy.getByDataTest('banner');
+    },
+    getBannerImage() {
+      return cy.getByDataTest('banner-image');
+    }
+  };
+}
+
+describe('testing Banner component', () => {
+  it('banner is render if the image loads', () => {
```
[the remark this answers is not in the document]
Comment (christianpe98) [reply packages/x-components/tests/unit/banner.spec.ts]: done

File: packages/x-components/src/x-modules/search/components/banner.vue [hunk 6]
```diff
@@ -18,11 +26,25 @@
   import { Component, Prop } from 'vue-property-decorator';
   import { xComponentMixin } from '../../../components/x-component.mixin';
   import { searchXModule } from '../x-module';
+
   /**
+   * Supported events for the anchor.
+   *
+   * @public
+   */
+  interface AnchorEvents {
```
[the remark this answers is not in the document]
Comment (christianpe98) [reply packages/x-components/src/x-modules/search/components/banner.vue]: done

File: packages/x-components/src/x-modules/search/components/banner.vue [hunk 7]
```diff
@@ -46,6 +75,21 @@
...
     }
+
+    /**
+     * Computed property which returns events supported by the anchor.
+     *
+     * @returns Events supported by the anchor.
+     *
+     * @internal
+     */
+    protected get anchorEvents(): AnchorEvents {
```
[the remark this answers is not in the document]
Comment (christianpe98) [reply packages/x-components/src/x-modules/search/components/banner.vue]: done

File: packages/x-components/src/__stubs__/banners-stubs.factory.ts [hunk 8]
```diff
@@ -9,31 +9,40 @@ import { Banner } from '@empathyco/x-types';
...
+ * @param url - The banner URL.
  * @param position - The banner position (= row) inside the grid.
+ * @param image - The image of the banner.
  *
  * @returns The banner.
  *
  * @internal
  */
-export function createBannerStub(identifier: string, position?: number): Banner {
+export function createBannerStub(
```
[the remark this answers is not in the document]
Comment (christianpe98) [reply packages/x-components/src/__stubs__/banners-stubs.factory.ts]: done

**Review by CachedaCodes — APPROVED** [review]

**Review by tiborux — APPROVED** [review]

**Pull request merged**